kvnemesis: add Raft application assertions

erikgrinaker opened the pull request:
This adds basic kvnemesis assertions for Raft command application:

* A request is only applied once, at a single log index and lease applied index across all replicas (i.e. no double-applies or replays).

* Commands do not regress the Raft index/term, lease applied index, and closed timestamp (assuming no node restarts).

* All replicas apply the same commands in the same order at the same positions.

Resolves #115771.
Touches #114421.
Epic: none
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/116319)
<!-- Reviewable:end -->

pavelkalinnikov commented on pkg/kv/kvserver/raft_apply_asserter.go:
```diff
@@ -0,0 +1,89 @@
+// licenses/APL.txt.
+
+// TODO: put in a test package
+package kvserver
```
Consider putting this into the `kvserver/apply` package if dependencies allow. This effectively codifies the contract.

erikgrinaker replied:
Yeah, good idea.

pavelkalinnikov commented on pkg/kv/kvserver/raft_apply_asserter.go:
```diff
@@ -0,0 +1,89 @@
+// snapshot index as applied on that replica.
+func (a *RaftApplyAsserter) Applied(
+	cmdID kvserverbase.CmdIDKey,
+	entry raftpb.Entry,
```
Only `entry.Index` is used, consider replacing this argument with just an index.

erikgrinaker replied:
We use `entry` to output the actual command on failures, so that we have some chance of figuring out what happened.

pavelkalinnikov commented on pkg/kv/kvserver/raft_apply_asserter.go:
```diff
@@ -0,0 +1,89 @@
+//
+// TODO: we need to handle snapshots too, by treating all commands up to the
+// snapshot index as applied on that replica.
+func (a *RaftApplyAsserter) Applied(
```
nit: consider ordering the coordinate parameters from big to small, i.e. (cmdID, rangeID->replicaID->index)

pavelkalinnikov commented on pkg/kv/kvserver/raft_apply_asserter.go:
```diff
@@ -0,0 +1,89 @@
+// Copyright 2019 The Cockroach Authors.
```
2023

pav-kv reviewed (commented):
Looking good.

Do we have an idea what conditions increase probability of double-apply situations? We could add a new kvnemesis test with tweaked [probabilities](https://github.com/cockroachdb/cockroach/blob/b9f7fb60a0826988b87e5ff1b33b5e51618444cb/pkg/kv/kvnemesis/generator.go#L324-L418) for the things that we throw into the mix.

I'm guessing though that the most interesting conditions are chaos things like slow connections and down nodes. The more lease/leader moves and snapshots, the merrier.

pavelkalinnikov commented on pkg/kv/kvserver/raft_apply_asserter.go:
```diff
@@ -0,0 +1,89 @@
+	mu syncutil.Mutex
+	// appliedCmds is a map of applied command IDs to the range ID and Raft log
+	// index at which it was applied.
+	appliedCmds map[kvserverbase.CmdIDKey]*appliedCmd
```
These are "local" assertions though that can be checked by an individual replica, we do have them in the code.

But maybe there is some kind of "global" assertion that would be more powerful here.

pavelkalinnikov replied:
The kind of "global" assertion that becomes possible is: after the test is done, piece each rangeID log together by joining it across its replicas. Then check for LAI and CT regressions.

pavelkalinnikov replied:
Double-apply is not the only problem we should be catching. Replica inconsistency can also be caused if there is a single apply that should not have happened. A global join of all the applied commands for a range between all replicas can help spotting such an illegal apply. The global LAI and CT regression check partially detects these cases.

We could also check each lease epoch applies a contiguous span of commands.

erikgrinaker replied:
I agree that we should extend this to include more assertions, but I'm inclined to limit it to only double-applies for now to get something shipped. We can extend it later.

We can consider recording the full sequence of events and only asserting after the test completes, which is maybe more in line with what we want for other assertions. However, if we want to record the actual command (to ease debugging failures) we can end up storing a lot of data in memory. Maybe that's ok though.

pavelkalinnikov commented on pkg/kv/kvserver/raft_apply_asserter.go:
```diff
@@ -0,0 +1,89 @@
+			replicas: map[roachpb.ReplicaID]bool{replicaID: true},
+		}
+	} else if ac.rangeID != rangeID {
+		// Command ID was used for a different range, ignore it.
```
This effectively disables double-apply tracking for this command in one of the 2 colliding ranges, right?

Make this case visible? Maybe log to `t.Log` or something.

Although this has a very low probability, so maybe fine as is.

erikgrinaker replied:
I think it's unlikely enough that we can ignore it. The product of this probability and the probability of a double-apply is negligible.

erikgrinaker replied:
Ended up tracking commands by range ID.

pavelkalinnikov commented on pkg/kv/kvserver/apply/asserter.go:
```diff
@@ -0,0 +1,92 @@
+// to call multiple times for the same command, as is commonly done via
+// StoreTestingKnobs.TestingApplyCalledTwiceFilter.
+//
+// TODO: we need to handle snapshots too, by treating all commands up to the
+// snapshot index as applied on that replica.
```
The beauty of the approach here is the property: for every command that applies, at least one replica calls `Apply`. It's totally possible that N-1 other replicas will receive a snapshot and won't report this command, but there is always one who does. (consider documenting this)

So, in the end `appliedCmds[rangeID]` contains all the commands the range applied. We can also reconstruct the log of all applied commands if we sort it by index/LAI.

erikgrinaker replied:
Yep. Will add a comment when I add snapshothandling. We probably also want to track and assert entry commits as well, but this will do for now.

erikgrinaker commented:
Ok, I've added basic double-apply assertions, with snapshot handling. Does this seem reasonable for a first stab? Anything else we should be checking while we're at it?

erikgrinaker commented:
I guess we could add LAI and CT tracking while we're at it, even though we do already have some code assertions for it. I'll have to refresh my memory on the invariants there, but "don't regress" seems easy enough. I might add that afterwards.

pavelkalinnikov commented on pkg/kv/kvserver/apply/asserter.go:
```diff
@@ -0,0 +1,142 @@
+//
+// It assumes that entry application is durable, i.e. that nodes won't lose
+// applied state. This is not true for real CRDB nodes which can lose applied
+// state on restart, but is generally true in tests. If needed, we can loosen
```
This wouldn't be true for tests that restart nodes, so I'd drop "generally true in tests".

Can more clearly say that it's true while the CRDB process runs. And when it restarts, the state can be rolled back.

pavelkalinnikov commented on pkg/kv/kvserver/apply/asserter.go:
```diff
@@ -0,0 +1,142 @@
+
+	a.ensureRangeLocked(rangeID)
+
+	// Record the applied index.
+	a.appliedIndex[rangeID][replicaID] = index
```
Do we need to assert that `index` > `appliedIndex`, just like in the command `Apply` flow? Or snapshots can be applied retrospectively?

erikgrinaker replied:
Snapshots can regress the applied index. I don't know how likely it is in practice, but in principle there's nothing stopping a leader from replicating entries to a replica concurrently with an in-flight snapshot.

pavelkalinnikov commented on pkg/kv/kvserver/apply/asserter.go:
```diff
@@ -0,0 +1,142 @@
+	// Record the applied commands for this replica. A command must have applied
+	// on at least one replica before it can be included in a snapshot, so all
+	// relevant commands are guaranteed to be present in the map.
+	for _, cmd := range a.appliedCmds[rangeID] {
+		if cmd.logIndex <= index {
+			cmd.replicas[replicaID] = true
+		} else {
+			delete(cmd.replicas, replicaID)
+		}
+	}
```
Not following this section. When we apply a snapshot, there is no evidence that either of the commands has been applied, so shouldn't do `cmd.replicas[replicaID] = true`.

Also not following why `delete(cmd.replicas, replicaID)` for higher indices. Could you clarify? Does a snapshot "cancel" all previously applied commands? Should we allow that?

pavelkalinnikov replied:
I think we should allow this. Applying a snapshot is like applying many commands. It should always move the applied state forward.

If we allow applying snapshot at lower indices, this jeopardizes durability. Say, one index is applied at a majority of nodes. We send a snapshot that downgrades the applied index -> now some entries may be applied on less than a majority of replicas.

erikgrinaker replied:
> When we apply a snapshot, there is no evidence that either of the commands has been applied

That's what a snapshot means: a snapshot at index I contains the result of applying all transitions (commands) <= I. Specifically, we want to assert that we don't apply a log entry with a command that was already applied via snapshot.

> Does a snapshot "cancel" all previously applied commands?

Yes, if we apply a snapshot that regresses the applied index then it effectively un-applies those commands.

pavelkalinnikov replied:
> a snapshot at index I contains the result of applying all transitions (commands) <= I

This is correct, but there is no evidence that the concrete commands with the `CommandID`s that we track are the ones that were applied in this snapshot. We only have evidence for those commands that were reported through the `Apply` call.

For the same reason, when a snapshot skips over some indices, we return an "ambiguous error" to the client, because we don't know whether the commands were applied or overridden by some other commands with lower indices.

pavelkalinnikov replied:
> Specifically, we want to assert that we don't apply a log entry with a command that was already applied via snapshot.

This we should verify via LAI regression checks. We should never have a command applied at higher index than a snapshot, but lower LAI than the snapsot's.

erikgrinaker replied:
>If we allow applying snapshot at lower indices, this jeopardizes durability. Say, one index is applied at a majority of nodes. We send a snapshot that downgrades the applied index -> now some entries may be applied on less than a majority of replicas.

In principle, it's safe as long as a quorum has durable logs up to the commit index. Consider the case where the quorum is at commit index 10, and we have a lagging follower at applied index 2. We can send it two snapshots at index 5 and then index 3 (which regresses its applied index) without changing the safety properties of the cluster (it still has a quorum at commit index 10). If we lost either of the quorum nodes, we can catch the follower up using the logs of the remaining replicas (those entries have already undergone consensus and no longer need quorum).

I'm not sure about what happens in practice though. We only catch up followers whose log lags behind the leader's truncated log, and the leader won't truncate its log below what it's already applied, so by definition the snapshot must move the applied index forward. However, there may be races where multiple snapshots are sent to a follower (e.g. with delegated snapshots), and an earlier snapshot replaces a later snapshot.

I guess we could assert that snapshots don't regress the applied index and see if something trips it. But I don't think it's an invariant we can rely on.

>This is correct, but there is no evidence that the concrete commands with the CommandIDs that we track are the ones that were applied in this snapshot.

That's an assumption we'll have to make here, unless we want to track the actual state at each log index. I'm not going to spend time on that right now. The consistency checker will detect this, and we run a consistency check after every kvnemesis run.

pavelkalinnikov replied:
> Consider the case where the quorum is at commit index 10, and we have a lagging follower at applied index 2. We can send it two snapshots at index 5 and then index 3 (which regresses its applied index) without changing the safety properties of the cluster (it still has a quorum at commit index 10).

I mean that we can't apply a snapshot that downgrades the **applied** index. Surely there can be snapshots below the committed index, but applying them is fine because in theory we can keep the corresponding suffix of the committed log (I don't know if it's true in practice, or we just wipe the log entirely when we apply a snapshot).

The durability of an entry on a replica may exist in either of the two (or both) ways:
- The entry is in the log.
- The entry is in the snapshot.

Once an entry is durable on a replica in either of the two ways, we can't revert it because raft relies on it for quorum purposes.

> I guess we could assert that snapshots don't regress the applied index and see if something trips it. But I don't think it's an invariant we can rely on.

I think we should rely on it. When we regress the applied index and there is no corresponding entry in the log (which likely can happen because the log can be truncated up to the applied index), we lost durability of this entry on this node.

> That's an assumption we'll have to make here, unless we want to track the actual state at each log index.

I disagree. We don't even make this assumption in the prod code. We simply don't know which entries are behind the applied snapshot. In the test-only code we should make strictly fewer assumptions.

I think we should record only the evidence, and then check that the evidence is self-consistent.

erikgrinaker replied:
Yeah, we probably can, because etcd/raft already asserts it (unless I'm missing something):

https://github.com/etcd-io/raft/blob/ee0fe9da492888b55fe183cf1a42931ad551ec6b/raft.go#L718-L720

https://github.com/etcd-io/raft/blob/1df762940b8c309a27cfafb086d767c0c7e3f58f/log.go#L329-L331

>I disagree. We don't even make this assumption in the prod code. We simply don't know which entries are behind the applied snapshot. In the test-only code we should make strictly fewer assumptions.

What do you mean? The applied state at any given applied index must be identical across replicas. If a replica tells us that it successfully applied a command at index I (which we record here) then the snapshot must include that command to be correct. Sure, it would be nice to actually verify snapshots (which is expensive and I won't spend time on now), but that shouldn't stop us from asserting what we can assert -- that we don't re-apply a command that we've supposedly) already received via a snapshot. What do we have to gain from that?

erikgrinaker replied:
Sure, done.

pavelkalinnikov commented on pkg/kv/kvserver/apply/asserter.go:
```diff
@@ -0,0 +1,145 @@
+	a.ensureRangeLocked(rangeID)
+
+	// Record the applied index.
+	a.appliedIndex[rangeID][replicaID] = index
```
What we could/should verify for a snapshot is:
- There is at least one applied command at the same index (i.e. someone has applied a command, and then sent a snapshot including it).
- (exception: the initial snapshot index, I think it's 10 or something; need to dig it)
- The LAI (and maybe some other metadata like CT) of this command matches what the snapshot state says.

erikgrinaker replied:
I'll leave this for later -- we need to start getting some runtime in the nightlies to look for reapplication.

pavelkalinnikov commented:
One detail about the `kvserverbase.CmdIDKey`: reproposals that bump the LAI, are proposed under a different ID (after Tobi's refactoring). Previously they all used to share the same ID, and the corresponding [proposals](https://github.com/cockroachdb/cockroach/blob/7ff6883d2e30a6571427a0f35441952ad4c4275f/pkg/kv/kvserver/replica.go#L713) map slot was reused for all of them.

To check at-most-once semantics for the entire proposal (including all its reproposals of both kinds), we need a canonical ID that is shared between all sub-proposals. The current implementation will only check at-most-once application for each LAI reprpoposal, but not the entire group.

To fix that, we can introduce a canonical uber-proposal ID that all sub-proposal copy around and share. Or, with the current code, you could use the ID of the [seedProposal](https://github.com/cockroachdb/cockroach/blob/44f13c7f48d4e703e2c8c1b9923cf0fa752f6db9/pkg/kv/kvserver/replica_proposal.go#L204-L206) as the canonical ID.

erikgrinaker committed:
kvserverbase: extend `ApplyFilterArgs`

Epic: none
Release note: None

erikgrinaker committed:
kvserverbase: extend `ProposalFilterArgs`

Epic: none
Release note: None

erikgrinaker committed:
kvserver: use `ProposalFilterArgs` for `TestingProposalSubmitFilter`

Epic: none
Release note: None

erikgrinaker committed:
kvserver: add `AfterSnapshotApplication` testing knob

Epic: none
Release note: None

erikgrinaker committed:
kvnemesis: don't randomly assign LAI for lease requests

Epic: none
Release note: None

erikgrinaker committed:
kvnemesis: add Raft application assertions

This adds basic kvnemesis assertions for Raft command application:

* A request is only applied once, at a single log index and lease
  applied index across all replicas (i.e. no double-applies or replays).

* Commands do not regress the Raft index/term, lease applied index, and
  closed timestamp (assuming no node restarts).

* All replicas apply the same commands in the same order at the same
  positions.

Epic: none
Release note: None

pavelkalinnikov commented on pkg/kv/kvserver/apply/asserter.go:
```diff
@@ -0,0 +1,308 @@
+//     do not regress (assuming no node restarts).
+type Asserter struct {
+	mu syncutil.Mutex
+	// proposedCmds tracks proposed commands by ID, along with the proposer.
+	// Reproposals under new command IDs are considered separate proposals here.
+	proposedCmds map[roachpb.RangeID]map[kvserverbase.CmdIDKey]roachpb.ReplicaID
+	// reproposedCmds keeps track of LAI reproposals under different command IDs
+	// (when the LAI changes), keyed by the original seed proposal's command ID
+	// and the reproposal's command ID.
+	reproposedCmds map[roachpb.RangeID]map[kvserverbase.CmdIDKey]map[kvserverbase.CmdIDKey]bool
+	// seedCmds maps a LAI reproposal command ID back to its original seed proposal.
+	// It is the reverse index of reproposedCmds.
+	seedCmds map[roachpb.RangeID]map[kvserverbase.CmdIDKey]kvserverbase.CmdIDKey
```
Is this level of detail needed here?

The contract that we're trying to verify is that each command is applied at most once.

Suppose we have a canonical command ID (which we do: it's `seedProposal`'s command ID; or the command's own ID if it is the seed). Then we just need to make sure that this ID is mentioned only once at exactly one log index etc.

Reproposals are an implementation detail that's not part of the visible behaviour. A command can be proposed, committed, and [almost] applied many times, but ultimately will apply only once.

```
command: proposal(LAI_1) -> reproposal(LAI_2) -> ... reproposal(LAI_k)
             |                    |                    |
             v                    v                    v
      reproposal(LAI_1)_1   reproposal(LAI_2)_1  ...  ...
             |                    |                    |
             v                    v                    v
      reproposal(LAI_1)_2   reproposal(LAI_2)_2.     reproposal(LAI_k)_m
          ...

             ^                    ^                    ^
             |                    |                    |
      all these are either never committed or     only one of these reproposals applies
      apply as no-ops (so Apply is not called)
```

If the apply contract works, the `Apply` function will ever be called only on one of these reproposals. If it's called on another one, this is a violation so we stop immediately.

Thus, it seems like we don't need to track the reproposals concept in the first place? Just the first occurrence.

pavelkalinnikov replied:
I think this is a balance. One inclination is to check as many invariants as possible. Another extreme is check only the things that this harness makes possible (which were impossible by other means): "global" invariants / safety properties. A bunch of per-replica/per-proposal properties can probably be verified inline as runtime assertions. This harness should mainly focus on what's not trivial or not possible locally.

---

One problem with the `seedProposal` though is that it's populated only on the leaseholder. Followers don't know if proposals with different IDs are related. This means that when the leaseholder `Apply` a reproposal, we will see a non-empty `seedProposal` etc; when a follower applies the same reproposal, we won't see it.


A snapshot applied on a leaseholder can render in skipping a few leaseholder-side `Apply` calls that carry the `seedProposal` signal. I.e. the ordering [follower-side `Apply` -> snapshot at leaseholder that covers this applied entry -> missing leaseholder-side `Apply`] introduces a gap in the double-apply detection.

To have 100% coverage in double-apply detection, we should either:
- (in this PR) intercept proposals on the leaseholder at stages earlier than apply, to memoize the ID->seedID mapping; then use this mapping to canonicalize the command IDs that come through the `Apply` call
- or (preferrably, longer-term) go back to having a command ID that is properly reused across reproposals.

erikgrinaker replied:
>One problem with the seedProposal though is that it's populated only on the leaseholder. Followers don't know if proposals with different IDs are related. This means that when the leaseholder Apply a reproposal, we will see a non-empty seedProposal etc; when a follower applies the same reproposal, we won't see it.

We record the seed/reproposal mapping during `Propose` (where we always know the seed), and recover it in `Apply`  regardless of which replica it's applying on.

>A bunch of per-replica/per-proposal properties can probably be verified inline as runtime assertions. This harness should mainly focus on what's not trivial or not possible locally.

Yes, we should add runtime assertions in the actual code as well, and that should be preferred. For now, I'm just trying to get some basic coverage in for #114421. You should feel free to make whatever changes you want later.

pavelkalinnikov replied:
Nice, having a `Propose`+`Apply` pair of trackers (for this and other reasons) makes sense.

erikgrinaker replied:
Rewrote the logic to track the command ID a seed command eventually applies as.

nvb reviewed (commented):
> Reviewed 1 of 7 files at r1, 13 of 13 files at r6, 2 of 2 files at r7, 3 of 3 files at r8, 2 of 2 files at r9, 1 of 1 files at r10, 6 of 7 files at r11, 1 of 1 files at r12, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/116319)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker and @pav-kv)

___
*[`pkg/kv/kvserver/apply/asserter.go` line 29 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/116319#-Nm3bxo-CcoS3T9QsXo9:-Nm3bxo-CcoS3T9QsXoA:b-66y9cc) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e8f24b0bd854edb5aaf65615e82c8d3d8b81ae7/pkg/kv/kvserver/apply/asserter.go#L29)):*
> ```Go
> //
> //   - A request is only applied once, at a single log index and lease applied index
> //     across all replica (i.e. no double-applies or replays).
> ```

Did you consider adding validation that a request is not skipped on some replicas, if it applied on other replicas?
___
*[`pkg/kv/kvserver/apply/asserter.go` line 37 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/116319#-Nm3XpRD6OAlLtIbsxRe:-Nm3XpRD6OAlLtIbsxRf:b-5x5izu) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e8f24b0bd854edb5aaf65615e82c8d3d8b81ae7/pkg/kv/kvserver/apply/asserter.go#L37)):*
> ```Go
> 	// proposedCmds tracks proposed commands by ID, along with the proposer.
> 	// Reproposals under new command IDs are considered separate proposals here.
> 	proposedCmds map[roachpb.RangeID]map[kvserverbase.CmdIDKey]roachpb.ReplicaID
> ```

We have 5 hash-maps here that are all keyed on rangeID, each with one or more inner hash-maps as values. It's all a bit difficult to read. Consider at least splitting the top-level map off into a separate structure. Something like:
```go
type Asserter struct {
    mu syncutil.Mutex
    ranges map[roachpb.RangeID]rangeAsserter
}

type rangeAsserter struct {
    proposedCmds map[kvserverbase.CmdIDKey]roachpb.ReplicaID
    // ...
}
```
___
*[`pkg/kv/kvserver/apply/asserter.go` line 45 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/116319#-Nm3YbDe6tVd7OlU9-Dr:-Nm3YbDe6tVd7OlU9-Ds:bwnbso9) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e8f24b0bd854edb5aaf65615e82c8d3d8b81ae7/pkg/kv/kvserver/apply/asserter.go#L45)):*
> ```Go
> 	// It is the reverse index of reproposedCmds.
> 	seedCmds map[roachpb.RangeID]map[kvserverbase.CmdIDKey]kvserverbase.CmdIDKey
> 	// appliedCmds tracks the log index at which a command is applied. Reproposals
> ```

s/log index/applied position/
___
*[`pkg/kv/kvserver/apply/asserter.go` line 196 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/116319#-Nm3ZgAx6fDNo9D-zn0N:-Nm3ZgAx6fDNo9D-zn0O:bmhjgdq) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e8f24b0bd854edb5aaf65615e82c8d3d8b81ae7/pkg/kv/kvserver/apply/asserter.go#L196)):*
> ```Go
> 	newReplicaApplied := pos
> 	if applied, ok := a.replicaApplied[rangeID][replicaID]; ok {
> 		// INVARIANT: the applied index must progress.
> ```

Could you leave a note about why this can have gaps. And consider also clarifying on the method signature that "apply" is in reference to "apply successfully without hitting the lease index check and being rejected".
___
*[`pkg/kv/kvserver/apply/asserter.go` line 249 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/116319#-Nm3aHmj13nodb50DuF1:-Nm3aHmj13nodb50DuF2:bgrwt8t) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e8f24b0bd854edb5aaf65615e82c8d3d8b81ae7/pkg/kv/kvserver/apply/asserter.go#L249)):*
> ```Go
> 		seedID = cmdID
> 	}
> 	otherCmds := map[kvserverbase.CmdIDKey]bool{seedID: true}
> ```

This logic could make it clearer what exactly we're trying to check and how we're going to check it, both in the comment and in the name of this variable. Consider `seedAndReproposedCmds` for the name.


<!-- Sent from Reviewable.io -->

erikgrinaker reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/116319)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten and @pav-kv)

___
*[`pkg/kv/kvserver/apply/asserter.go` line 29 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/116319#-Nm3bxo-CcoS3T9QsXo9:-Nm61ZNm8xCzDSwuqRsC:b-nxagg6) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e8f24b0bd854edb5aaf65615e82c8d3d8b81ae7/pkg/kv/kvserver/apply/asserter.go#L29)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Did you consider adding validation that a request is not skipped on some replicas, if it applied on other replicas?
</blockquote></details>

I dropped it in the interest of time -- I was hoping to have something simple merged last week, so we could start to get some coverage for #114421. I have other things I need to wrap up before the holidays, so I'd prefer to get this in as-is and extend the assertions later, rather than letting this sit unmerged for the next couple of weeks.
___
*[`pkg/kv/kvserver/apply/asserter.go` line 37 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/116319#-Nm3XpRD6OAlLtIbsxRe:-Nm5rQS24ijZl_C3J-5h:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e8f24b0bd854edb5aaf65615e82c8d3d8b81ae7/pkg/kv/kvserver/apply/asserter.go#L37)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

We have 5 hash-maps here that are all keyed on rangeID, each with one or more inner hash-maps as values. It's all a bit difficult to read. Consider at least splitting the top-level map off into a separate structure. Something like:
```go
type Asserter struct {
    mu syncutil.Mutex
    ranges map[roachpb.RangeID]rangeAsserter
}

type rangeAsserter struct {
    proposedCmds map[kvserverbase.CmdIDKey]roachpb.ReplicaID
    // ...
}
```
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/apply/asserter.go` line 249 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/116319#-Nm3aHmj13nodb50DuF1:-Nm636VYAENZaZ8feToI:b-u7oy7i) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e8f24b0bd854edb5aaf65615e82c8d3d8b81ae7/pkg/kv/kvserver/apply/asserter.go#L249)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

This logic could make it clearer what exactly we're trying to check and how we're going to check it, both in the comment and in the name of this variable. Consider `seedAndReproposedCmds` for the name.
</blockquote></details>

Rewrote the logic.


<!-- Sent from Reviewable.io -->

erikgrinaker reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/116319)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten and @pav-kv)

___
*[`pkg/kv/kvserver/apply/asserter.go` line 29 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/116319#-Nm3bxo-CcoS3T9QsXo9:-NmBK5Tv5-cvMjsCZh8n:b-1h3xy) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e8f24b0bd854edb5aaf65615e82c8d3d8b81ae7/pkg/kv/kvserver/apply/asserter.go#L29)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

I dropped it in the interest of time -- I was hoping to have something simple merged last week, so we could start to get some coverage for #114421. I have other things I need to wrap up before the holidays, so I'd prefer to get this in as-is and extend the assertions later, rather than letting this sit unmerged for the next couple of weeks.
</blockquote></details>

Rewrote this to track the applied Raft log, and added additional assertions, including that all replicas apply all commands.

Hoping we can get this merged today, before the holidays, since we need to get some runtime before shipping 23.2.


<!-- Sent from Reviewable.io -->

pav-kv reviewed (approved):
Seems good to start with. I didn't review too closely, good if Nathan takes another look.

erikgrinaker commented on pkg/kv/kvserver/apply/asserter.go:
```diff
@@ -0,0 +1,411 @@
+	}
+
+	if req.IsSingleRequestLeaseRequest() {
+		// INVARIANT: lease requests never set a LAI or closed timestamp. These have
+		// their own replay protection.
```
It's a conditional write, predicated on the old lease record. Mentioned it in the comment.

erikgrinaker commented on pkg/kv/kvserver/apply/asserter.go:
```diff
@@ -0,0 +1,411 @@
+		// INVARIANT: entries may not regress the applied state.
+		if len(r.log) > 0 {
+			s := r.stateAt(kvpb.RaftIndex(len(r.log) - 1))
+			// INVARIANT: the Raft index must progress.
+			if entry.raftIndex <= s.raftIndex {
+				fail("Raft index regression %d -> %d", s.raftIndex, entry.raftIndex)
+			}
```
It is, but figured we can keep it in as documentation anyway. Added a comment saying as much.

erikgrinaker commented on pkg/kv/kvserver/apply/asserter.go:
```diff
@@ -0,0 +1,411 @@
+			if entry.raftIndex <= s.raftIndex {
+				fail("Raft index regression %d -> %d", s.raftIndex, entry.raftIndex)
+			}
+			// INVARIANT: the Raft term must not regress.
+			if entry.raftTerm < s.raftTerm {
+				fail("Raft term regression %d -> %d", s.raftTerm, entry.raftTerm)
+			}
```
Yep, but that's the case for many other assertions here as well (e.g. that all replicas apply the same command at the same index). We may as well assert these invariants here as well, since we rely on them holding. Don't feel particularly strongly about it though, but it doesn't really cost us anything either.

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 2 files at r21, 2 of 3 files at r22, 2 of 2 files at r23, 4 of 4 files at r25, 1 of 1 files at r26, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/116319#-:-Nn9lvYg8PDfOL7j8bjN:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @erikgrinaker and @pav-kv)



<!-- Sent from Reviewable.io -->

erikgrinaker commented:
TFTR!

bors r+
